rfcs/0011-separate-flyway-from-pipeline.md

```diff
@@ -0,0 +1,382 @@
+deploy a schema change separately from its accompanying application change or have forward-compatible changes (since the
+new application will only ever start up after the new schema has been deployed). Backward-compatibility remains a
+requirement because the old applications will still coexist with the new schema for a brief period of time.
+
+Moving Hibernate validation out of the BFD Pipeline and BFD FHIR Server avoids the issue of old applications coming up
+during an auto-scaling event and failing Hibernate validation.
+
+Long-running migrations will no longer block the BFD Pipeline application processing because the old pipeline
+application will continue to run and process data until the schema migration completes.
+
+The new step that runs the Flyway migrations will be configured to run as a user that has the appropriate privileges to
+run migrations without having to elevate the privileges of the BFD Pipeline. This will allow schema migrations that
+create (non-user) database roles and grant privileges to run in Flyway.
+
+### Proposed Solution: Detailed Design
+[Proposed Solution: Detailed Design]: #proposed-solution-detailed-design
+
+Application design:
+
+* A new Java application 'bfd-db-migrator' will be introduced that performs these functions:
+  * Invokes Flyway (and thereby runs all pending migrations)
+  * Runs Hibernate validation against the BFD ORM (after Flyway finishes)
+  * Exits with a return code that indicates whether all operations were successful or not
+
+* Flyway invocations and hibernate validation will be removed from the BFD Pipeline and BFD API Server applications.
+
+* Test infrastructure will be altered to mimic the deployment by running the new application prior to starting the local
+BFD server for tests that run the BFD server as a standalone process.
+
+* Test infrastructure will be altered to invoke the new application logic via method call prior to running tests that
+require a BFD database but do not launch a BFD server.
+
+* The new Java application code will reside in the existing BFD repo as another sub-module of bfd-parent so that it will
+be built and tested as part of the current Github actions application workflow and Jenkins App build stage.
+
+Deployment design:
+
+![Proposed Deployment Sequence](./resources/0011-proposed-deployment-sequence.svg)
+
+<details><summary>Diagram Notes</summary>
+Like the previous diagram, this illustrates the deployment sequence with some broad definitions. This adds the `bfd-db-migrator` participant that includes a blocking deployment stage for each environment that must complete **before** proceeding to apply updates to the `bfd-pipeline` and `bfd-server` resources for each environment.
+</details>
+
+* A new migrator service account and associated migrator database role that has all required privileges for running
+migrations and hibernate validation will be created.
```

Comment:
Moderate question: Do we know if this ends up being the DB owner?

Reply:
I'm not 100% sure but that was my assumption. We should verify this as part of the implementation. I will add it to the open questions.